### ISSUE TITLE
SVI leaking memory?

### ISSUE THREAD

**Issue by nh2liu** [pyro-ppl/numpyro#622]
I'm having issues with numpyro, specifically `SVI` using GPU memory when running the same model multiple times.

I've set `XLA_PYTHON_CLIENT_ALLOCATOR=platform` as advised in jax docs to debug this further.

Here's a sample gist of the code I am using to debug this.

https://gist.github.com/nh2liu/2570812c5913a8e0889e746274188aba

The loop at the bottom runs a SVI for each iteration. Even though nothing is returned (python should dereference and GC), I'm seeing the GPU memory usage increasing on `nvidia-smi` and this script will eventually throw a jax out of memory when trying to allocate exception.

I've also tested this with a sample normal function with jax and did not see this behavior.

**Comment by neerajprad** [comment]
Thanks for the sample gist, @nh2liu. Do you also see this with the CPU? The only thing that I can think of is that JAX has a compilation cache which can grow without bound. I don't have a GPU available currently, but could you check if the number of `jax.interpreters.xla.DeviceArray` objects increases as the loop runs (using `gc.get_objects`)?

**Comment by fehiepsi** [comment]
> JAX has a compilation cache which can grow without bound

Agreed! I think JAX users usually solve this issue by doing e.g. `jit(train)(X, y, key)`.

**Comment by fehiepsi** [comment]
Yeah! Actually, it will rarely affect functional programmers who don't use `lambda`. :D
